types/react-table/index.d.ts

```diff
@@ -1,746 +1,673 @@
+    prepareRow: Array<(row: Row<D>, instance: TableInstance<D>) => Row<D>>;
+    getTableProps: Array<(props: object, instance: TableInstance<D>) => object>;
+    getTableBodyProps: Array<(props: object, instance: TableInstance<D>) => object>;
+    getHeaderGroupProps: Array<(props: object, instance: TableInstance<D>, header: HeaderGroup<D>) => object>;
+    getFooterGroupProps: Array<(props: object, instance: TableInstance<D>, header: HeaderGroup<D>) => object>;
+    getHeaderProps: Array<(props: object, instance: TableInstance<D>, header: HeaderGroup<D>) => object>;
+    getFooterProps: Array<(pops: object, instance: TableInstance<D>, header: HeaderGroup<D>) => object>;
+    getRowProps: Array<(props: object, instance: TableInstance<D>, row: Row<D>) => object>;
+    getCellProps: Array<(props: object, instance: TableInstance<D>, cell: Cell<D>) => object>;
+    useFinalInstance: Array<(instance: TableInstance<D>) => void>;
+}
+
+export interface UseTableColumnOptions<D extends object>
+    extends Accessor<D>,
+        Partial<{
+            columns: Array<Column<D>>;
+            show: boolean | ((instance: TableInstance<D>) => boolean);
+            Header: Renderer<HeaderProps<D>>;
+            Cell: Renderer<CellProps<D>>;
+            width?: number | string;
+            minWidth?: number;
+            maxWidth?: number;
+        }> {}
+
+type UpdateHiddenColumns<D extends object> = (oldHidden: Array<IdType<D>>) => Array<IdType<D>>;
+
+export interface UseTableInstanceProps<D extends object> {
+    state: TableState<D>;
+    hooks: Hooks<D>;
+    dispatch: TableDispatch;
+    columns: Array<ColumnInstance<D>>;
+    flatColumns: Array<ColumnInstance<D>>;
+    headerGroups: Array<HeaderGroup<D>>;
+    footerGroups: Array<HeaderGroup<D>>;
+    headers: Array<ColumnInstance<D>>;
+    flatHeaders: Array<ColumnInstance<D>>;
+    rows: Array<Row<D>>;
+    getTableProps: (props?: object) => object;
+    getTableBodyProps: (props?: object) => object;
+    prepareRow: (row: Row<D>) => void;
+    flatRows: Array<Row<D>>;
+    totalColumnsWidth: number;
+    toggleHideColumn: (columnId: IdType<D>, value?: boolean) => void;
+    setColumnHidden: (param: Array<IdType<D>> | UpdateHiddenColumns<D>) => void;
+    toggleHideAllColumns: (value?: boolean) => void;
+    getToggleHideAllColumnsProps: (userProps: any) => any;
+}
+
+export interface UseTableHeaderGroupProps<D extends object> {
+    headers: Array<ColumnInstance<D>>;
+    getHeaderGroupProps: (props?: object) => object;
+    getFooterGroupProps: (props?: object) => object;
+    totalHeaderCount: number; // not documented
 }
 
-export interface SortingRule {
+export interface UseTableColumnProps<D extends object> {
+    id: IdType<D>;
+    columns: Array<ColumnInstance<D>>;
+    isVisible: boolean;
+    render: (type: 'Header' | 'Footer' | string, props?: object) => ReactNode;
+    totalLeft: number;
+    totalWidth: number;
+    getHeaderProps: (props?: object) => object;
```

Comment:
for prop getters like this, can we type the response as `object & { key: string }` in lieu of a more comprehensive interface?

Reply:
I think we could maybe declare separate types outside, one for each prop getter, like `TableHeaderProps`, `TableRowProps`, `TableCellProps` etc. and then use those in prop getters? Like

```ts
export interface TableCellProps {
  key?: React.Key;
}

// …

getCellProps: (props?: TableCellProps) => TableCellProps;
```

This would allow us to use declaration merging to extend those interfaces easily, adding the props that we would be using?

Reply:
Yeah absolutely. I just thought it was worth adding `key` at the very least since that needs to passed explicitly if you're enforcing `react/jsx-key` in your eslint config.

Reply:
Ah, yes, hmm, it would be nice to make the key mandatory in the output, but optional in the input!

Reply:
done, separate types for each one.